File: week1/BOJ_7576_토마토/seokjun_7576.js
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

class Queue {
  constructor() {
    this.items = [];
    this.front = 0;
    this.rear = 0;
  }

  enqueue(element) {
    this.items.push(element);
    this.rear += 1;
  }

  dequeue() {
    return this.items[this.front++];
  }

  front() {
    return this.items[front];
  }

  isEmpty() {
    return this.front === this.rear;
  }

  size() {
    return this.rear - this.front;
  }
}

const input = [];
let currentLine = 0;

rl.on("line", (line) => {
  input.push(line.split(" ").map(Number));
});

rl.on("close", () => {
  const [M, N] = input[currentLine]; // M은 가로길이, N은 세로 길이

  const box = Array.from({ length: N }, () => []);

  for (let n = 0; n < N; n++) {
    currentLine += 1;
    for (let m = 0; m < M; m++) {
      box[n].push(input[currentLine][m]);
    }
  }

  const dirs = [1, -1, 0, 0, 1, -1];
  const bfs = (maturedPositions) => {
    const q = new Queue();

    for (const matured of maturedPositions) {
      q.enqueue(matured);
    }

    while (q.size() > 0) {
      [y, x] = q.dequeue();
      let day = box[y][x];

      for (let k = 0; k < 4; k++) {
        newY = y + dirs[k];
        newX = x + dirs[k + 2];
        if (
          newY < 0 ||
          newX < 0 ||
          newY >= N ||
          newX >= M ||
          box[newY][newX] !== 0
        ) {
          continue;
        }
        q.enqueue([newY, newX]);
        box[newY][newX] = day + 1;
      }
    }
  };

  const maturedPositions = [];
  for (let i = 0; i < N; i++) {
    for (let j = 0; j < M; j++) {
      if (box[i][j] === 1) {
        maturedPositions.push([i, j]);
      }
    }
  }
  bfs(maturedPositions);

  const getDays = () => {
    let max = -Infinity;
    for (let i = 0; i < N; i++) {
      for (let j = 0; j < M; j++) {
        if (box[i][j] === 0) {
          return -1;
        }

        if (box[i][j] > max) {
          max = box[i][j];
        }
      }
    }

    return max - 1;
  };

  console.log(getDays());
  process.exit();
});

File: week1/BOJ_7576_토마토/seokjun_7576.py
from collections import deque

M, N = map(int, input().split(" ")) # M은 가로길이, N은 세로길이
box = [[] for _ in range(N)]

for n in range(N):
  row = list(map(int, input().split(" ")))
  for m in range(M):
    box[n].append(row[m])

dirs = [1, -1, 0, 0, 1, -1]
def bfs(maturedPositions):
  q = deque()

  for matured in maturedPositions:
    q.append(matured)

  while q:
    position = q.popleft()
    y = position[0]
    x = position[1]
    day = box[y][x]

    for i in range(4):
      newY = y + dirs[i]
      newX = x + dirs[i + 2]

      if newY < 0 or newX < 0 or newY >= N or newX >= M or box[newY][newX] != 0:
        continue

      q.append((newY, newX))
      box[newY][newX] = day + 1

maturedPositions = []
for y in range(N):
  for x in range(M):
    if box[y][x] == 1:
      maturedPositions.append((y, x))

bfs(maturedPositions)

def getDays():
  max = -1
  for y in range(N):
    for x in range(M):
      if box[y][x] == 0:
        return -1

      if box[y][x] > max:
        max = box[y][x]

  return max - 1

print(getDays())